### PR TITLE
CASSANDRA-16082 - Add a new jmxtool which can dump what JMX objects exist and diff

### PR DESCRIPTION
**Pull request by dcapwell**

### CODE REVIEW

File: test/unit/org/apache/cassandra/tools/ToolRunner.java [hunk 1]
```diff
@@ -75,13 +77,6 @@ public ToolRunner(List<String> args, boolean runOutOfProcess)
         this.runOutOfProcess = runOutOfProcess;
     }
 
-    public ToolRunner withStdin(InputStream stdin, boolean autoClose)
```
Comment (bereng) [inline test/unit/org/apache/cassandra/tools/ToolRunner.java:78]: Why kill stdin?

File: test/unit/org/apache/cassandra/tools/ToolRunner.java [hunk 2]
```diff
@@ -304,17 +248,12 @@ public boolean isRunning()
     }
 
     public boolean waitFor()
-    {
-        return waitFor(defaultTimeoutMillis, TimeUnit.MILLISECONDS);
-    }
-
-    public boolean waitFor(long time, TimeUnit timeUnit)
```
Comment (bereng) [inline test/unit/org/apache/cassandra/tools/ToolRunner.java:311]: Why kill this method? and if it is so why not remove `defaultTimeoutMillis` along?

File: test/unit/org/apache/cassandra/tools/ToolRunner.java [hunk 3]
```diff
@@ -425,14 +364,56 @@ public void close()
...
+
+        private StreamGobbler(InputStream input, CompletableFuture<ByteArrayOutputStream> whenComplete)
+        {
+            this.input = input;
+            this.whenComplete = whenComplete;
+        }
+
+        public void run()
+        {
+            byte[] buffer = new byte[1 << 13];
```
Comment (bereng) [inline test/unit/org/apache/cassandra/tools/ToolRunner.java]: Replace `1 << 13` imo for a human readable equivalent of 8K as it's a constant and you're not in a hot path with a variable either.

File: test/unit/org/apache/cassandra/tools/ToolRunner.java (hunk 1 above)
Comment (dcapwell) [reply test/unit/org/apache/cassandra/tools/ToolRunner.java:78]: it was dead code and harder to reason how to deal with.  There are two patterns I see with stdin: one time input, and interactive; the interactive case seems hard to deal with if creating a thread to try to copy bytes.  Normally I see stdin exposed so the test can control when bytes are added, so would expect something like

```class ToolRunner {
...
OutputStream getStdin()
...```. and then the test can append to stdin when desired.

Comment (dcapwell) [reply test/unit/org/apache/cassandra/tools/ToolRunner.java:78]: Then again, we don't stream stdout so the interactive case isn't possible currently (do we even have that with any tool?)

File: test/unit/org/apache/cassandra/tools/ToolRunner.java (hunk 2 above)
Comment (dcapwell) [reply test/unit/org/apache/cassandra/tools/ToolRunner.java:311]: It causes subtle bugs since the API doesn't handle the case the process has not completed.

The APIs check for exit code and will call waitFor, but if waitFor times out then the exit code stays 0 and the test passes with a bad result.  Rather than trying to change the API to spilt these two cases, I just removed the timeout waitFor since I couldn't find a single usage which properly handled timeout; all usages assumed no timeout and would move forward incorrectly when a timeout happens.

File: test/unit/org/apache/cassandra/tools/ToolRunner.java (hunk 3 above)
Comment (dcapwell) [reply test/unit/org/apache/cassandra/tools/ToolRunner.java]: `1 << 13` is a compile time constant, so javac compiles it down to 8k

```
$ javap -v -p -cp build/classes/main/:build/test/classes/ 'org.apache.cassandra.tools.ToolRunner$StreamGobbler'
...
public void run();
    descriptor: ()V
    flags: ACC_PUBLIC
    Code:
      stack=4, locals=3, args_size=1
         0: sipush        8192
         3: newarray       byte
```

I will change to a constant.

File: test/unit/org/apache/cassandra/tools/ToolRunner.java (hunk 1 above)
Comment (dcapwell) [reply test/unit/org/apache/cassandra/tools/ToolRunner.java:78]: I will add it back for now.

File: test/unit/org/apache/cassandra/tools/ToolRunner.java [hunk 4]
```diff
@@ -196,52 +186,6 @@ public int waitFor() throws InterruptedException
...
-        while (true)
-        {
-            boolean errHandled;
-            boolean outHandled;
-            try
-            {
-                if (stdin != null)
-                {
-                    IOUtils.copy(stdin, in);
-                    if (stdinAutoClose)
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/tools/ToolRunner.java:213]: In trying to add stdin support back bound that this logic doesn't make sense.  `IOUtils.copy` blocks waiting for the stream to close, so doesn't make sense to close the closed stream.

File: test/unit/org/apache/cassandra/tools/ToolRunner.java [hunk 5]
```diff
@@ -196,52 +186,6 @@ public int waitFor() throws InterruptedException
...
-        InputStream out = process.getInputStream();
-        while (true)
-        {
-            boolean errHandled;
-            boolean outHandled;
-            try
-            {
-                if (stdin != null)
-                {
-                    IOUtils.copy(stdin, in);
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/tools/ToolRunner.java:212]: Also, this assumes the stream will one day close... to enforce that I am adding logic to try to close stdin when the process exists (else the thread blocks forever)

File: test/unit/org/apache/cassandra/tools/ToolRunner.java (hunk 1 above)
Comment (dcapwell) [reply test/unit/org/apache/cassandra/tools/ToolRunner.java:78]: I added stdin back but mostly had to redo the logic.  Since we don't control stdin it might not be closed when the process exists which would cause threads to leak (as copy would hang forever); to work around this I now attempt to detect when the process exists and run callbacks (only used to close stdin).

File: test/unit/org/apache/cassandra/tools/ToolRunner.java (hunk 2 above)
Comment (dcapwell) [reply test/unit/org/apache/cassandra/tools/ToolRunner.java:311]: removed default timeout

Comment (bereng) [reply test/unit/org/apache/cassandra/tools/ToolRunner.java:311]: Ok I see you did a first pass on CASSANDRA-15991 where it is being used. Let's remove it for the time now...

File: src/java/org/apache/cassandra/tools/JMXTool.java [hunk 6]
```diff
@@ -0,0 +1,795 @@
...
+public class JMXTool
+{
+    private static final List<String> METRIC_PACKAGES = Arrays.asList("org.apache.cassandra.metrics",
+                                                                      "org.apache.cassandra.db",
+                                                                      "org.apache.cassandra.hints",
+                                                                      "org.apache.cassandra.internal",
+                                                                      "org.apache.cassandra.net",
+                                                                      "org.apache.cassandra.request",
+                                                                      "org.apache.cassandra.service"
+    );
```
Comment (bereng) [inline src/java/org/apache/cassandra/tools/JMXTool.java]: nit: move bracket closing to upper line?

File: tools/bin/jmxtool.bat [hunk 7]
```diff
@@ -0,0 +1,36 @@
+@REM
```
Comment (bereng) [inline tools/bin/jmxtool.bat:1]: Where did we land with removing Windows support? iirc things were leaning towards actually removing it so maybe you could spare this .bat.

File: src/java/org/apache/cassandra/tools/JMXTool.java [hunk 8]
```diff
@@ -0,0 +1,795 @@
...
+
+        @Option(title = "exclude attributes", name = "--exclude-attribute", description = "Ignores processing specific attributes")
+        private List<CliPattern> excludeAttributes = new ArrayList<>();
+
+        @Option(title = "exclude operations", name = "--exclude-operation", description = "Ignores processing specific operations")
+        private List<CliPattern> excludeOperations = new ArrayList<>();
+
+        public Void call() throws Exception
+        {
+            if (files.size() != 2)
```
Comment (bereng) [inline src/java/org/apache/cassandra/tools/JMXTool.java]: Guava `Preconditions` are nice for these kind of checks #collaborating

File: test/unit/org/apache/cassandra/tools/JMXCompatabilityTest.java [hunk 9]
```diff
@@ -0,0 +1,178 @@
+package org.apache.cassandra.tools;
```
Comment (bereng) [inline test/unit/org/apache/cassandra/tools/JMXCompatabilityTest.java:1]: What about adding tests for the tool params, testing those themselves and making sure they indeed work? The different formats could be sthg to test i.e.

File: test/unit/org/apache/cassandra/tools/ToolRunner.java [hunk 10]
```diff
@@ -159,20 +162,7 @@ public InputStream getInputStream()
...
-                        {
-                            out = new ByteArrayOutputStream(stdin.available());
-                            IOUtils.copy(stdin, out);
-                        }
-                        catch(IOException e)
-                        {
-                            throw new RuntimeException("Failed to get stdin", e);
-                        }
-                        return out;
+                        return null;
```
Comment (bereng) [inline test/unit/org/apache/cassandra/tools/ToolRunner.java]: Doesn't this need to change now stdin is back in? Also don't you need to call `System.setIn()` somewhere?

File: tools/bin/jmxtool.bat (hunk 7 above)
Comment (dcapwell) [reply tools/bin/jmxtool.bat:1]: rather leave for now since all other tools have it

File: test/unit/org/apache/cassandra/tools/JMXCompatabilityTest.java (hunk 9 above)
Comment (dcapwell) [reply test/unit/org/apache/cassandra/tools/JMXCompatabilityTest.java:1]: you mean unit tests for the tool or the utilities in the tool?

Comment (bereng) [reply test/unit/org/apache/cassandra/tools/JMXCompatabilityTest.java:1]: You can test that in/out different formats are indeed working. Does specifying json _really_ produce a json file i.e?

Also wdty about adding a test case to detect help message changes, to remind you of changing docs as well as we have in 15991?

Comment (dcapwell) [reply test/unit/org/apache/cassandra/tools/JMXCompatabilityTest.java:1]: pushed tests to validate json -> json, yaml -> yaml work, but also added the help tests you are adding in the other one.

didn't add diff/dump tests as they are covered by the compatibility test.

File: test/unit/org/apache/cassandra/tools/ToolRunner.java (hunk 10 above)
Comment (dcapwell) [reply test/unit/org/apache/cassandra/tools/ToolRunner.java]: sorry, I only tested forking for stdin and neglected not forking =(

Comment (dcapwell) [reply test/unit/org/apache/cassandra/tools/ToolRunner.java]: added back

**Comment by dcapwell** [comment]
@bereng fixed or applied to all comments

File: tools/bin/jmxtool.bat (hunk 7 above)
Comment (bereng) [reply tools/bin/jmxtool.bat:1]: sgtm

File: src/java/org/apache/cassandra/io/util/DataOutputBuffer.java [hunk 11]
```diff
@@ -90,7 +90,7 @@ public DataOutputBuffer(ByteBuffer buffer)
     @Override
     public void flush() throws IOException
     {
-        throw new UnsupportedOperationException();
+
```
Comment (bereng) [inline src/java/org/apache/cassandra/io/util/DataOutputBuffer.java:93]: That worries me. Can you please fire a CI run? I would do it for you but I don't have the privileges.

Comment (dcapwell) [reply src/java/org/apache/cassandra/io/util/DataOutputBuffer.java:93]: Tests are https://app.circleci.com/pipelines/github/dcapwell/cassandra?branch=jmxDump, I can trigger Python dtests but unit/jvm-dtests are green.

File: test/unit/org/apache/cassandra/tools/ToolRunner.java [hunk 12]
```diff
@@ -176,17 +176,51 @@ public OutputStream getOutputStream()
...
+            {
+                ioWatchers[2] = new Thread(new StreamGobbler<>(stdin, process.getOutputStream()));
+                ioWatchers[2].setDaemon(true);
+                ioWatchers[2].setName("IO Watcher stdin for " + allArgs);
+                ioWatchers[2].start();
+                // since stdin might not close the thread would block, so add logic to try to close stdin when the process exits
+                onComplete.whenComplete((i1, i2) -> {
+                    try
+                    {
+                        stdin.close();
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/tools/ToolRunner.java:216]: sadly this is leaky as the life-cycle isn't clearly monitored so there are edge cases where this may start to leak...

File: src/java/org/apache/cassandra/io/util/DataOutputBuffer.java (hunk 11 above)
Comment (jonmeredith) [reply src/java/org/apache/cassandra/io/util/DataOutputBuffer.java:93]: I feel uncomfortable about removing this protection too.  Would it be straightforward to set some kind of system property to permit DataOutputBuffer.flush in the code that needs it?

File: tools/bin/jmxtool [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+  esac
+  shift
+done
+
+if [ "x$MAX_HEAP_SIZE" = "x" ]; then
+    MAX_HEAP_SIZE="512m"
+fi
+
+"$JAVA" $JAVA_AGENT -ea -da:net.openhft... -cp "$CLASSPATH" $JVM_OPTS -Xmx$MAX_HEAP_SIZE \
+        -Dlog4j.configurationFile=log4j2-tools.xml \
```
Comment (jonmeredith) [inline tools/bin/jmxtool]: Guessing should be `-Dlogback.configurationFile=logback-tools.xml`, is this a copy/paste from `fqltool`?

File: src/java/org/apache/cassandra/tools/JMXTool.java [hunk 14]
```diff
@@ -0,0 +1,838 @@
...
+                                                                      "org.apache.cassandra.request",
+                                                                      "org.apache.cassandra.service");
+
+    private static final Comparator<MBeanOperationInfo> OPERATOR_COMPARATOR = (a, b) -> {
+        int rc = a.getName().compareTo(b.getName());
+        if (rc != 0)
+            return rc;
+        String[] aSig = Stream.of(a.getSignature()).map(MBeanParameterInfo::getName).toArray(String[]::new);
+        String[] bSig = Stream.of(b.getSignature()).map(MBeanParameterInfo::getName).toArray(String[]::new);
+        rc = Integer.compare(aSig.length, bSig.length);
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/tools/JMXTool.java:80]: A shame you can't use `return Arrays.compare(aSig, bSig);` J9 and up :(

File: src/java/org/apache/cassandra/tools/JMXTool.java [hunk 15]
```diff
@@ -0,0 +1,838 @@
...
+    @Command(name = "dump", description = "Dump the Apache Cassandra JMX objects and metadata.")
+    public static final class Dump implements Callable<Void>
+    {
+        @Inject
+        private HelpOption helpOption;
+
+        @Option(title = "url", name = { "-u", "--url" }, description = "JMX url to target")
+        private String targetUrl = "service:jmx:rmi:///jndi/rmi://localhost:7199/jmxrmi";
+
+        @Option(title = "format", name = { "-f", "--format" }, description = "What format to dump content as")
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/tools/JMXTool.java]: Be kind to those users and let them know the values in `Format` so they don't have to dig in the source.

File: src/java/org/apache/cassandra/tools/JMXTool.java [hunk 16]
```diff
@@ -0,0 +1,838 @@
+package org.apache.cassandra.tools;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/tools/JMXTool.java]: unused

File: src/java/org/apache/cassandra/tools/JMXTool.java [hunk 17]
```diff
@@ -0,0 +1,838 @@
...
+import java.io.OutputStream;
+import java.io.OutputStreamWriter;
+import java.io.PrintStream;
+import java.io.Reader;
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/tools/JMXTool.java]: Unused

File: src/java/org/apache/cassandra/tools/JMXTool.java [hunk 18]
```diff
@@ -0,0 +1,838 @@
...
+import java.io.UnsupportedEncodingException;
+import java.nio.charset.StandardCharsets;
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/tools/JMXTool.java]: Unused

File: src/java/org/apache/cassandra/tools/JMXTool.java [hunk 19]
```diff
@@ -0,0 +1,838 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/tools/JMXTool.java]: Unused.

File: test/unit/org/apache/cassandra/tools/ToolRunner.java [hunk 20]
```diff
@@ -52,15 +50,17 @@
     protected static final Logger logger = LoggerFactory.getLogger(ToolRunner.class);
 
     private static final ImmutableList<String> DEFAULT_CLEANERS = ImmutableList.of("(?im)^picked up.*\\R");
-    
+    private static final String[] EMPTY_STRING_ARRAY = new String[0];
```
Comment (bereng) [inline test/unit/org/apache/cassandra/tools/ToolRunner.java:52]: Looks like this is only used once. Do you need to make it a class constant?

File: test/unit/org/apache/cassandra/tools/ToolRunner.java [hunk 21]
```diff
@@ -52,15 +50,17 @@
...
+
     private final List<String> allArgs = new ArrayList<>();
     private Process process;
+    @SuppressWarnings("resource")
     private final ByteArrayOutputStream errBuffer = new ByteArrayOutputStream();
+    @SuppressWarnings("resource")
     private final ByteArrayOutputStream outBuffer = new ByteArrayOutputStream();
+    private final CompletableFuture<Void> onComplete = new CompletableFuture<>();
```
Comment (bereng) [inline test/unit/org/apache/cassandra/tools/ToolRunner.java]: This looks unnecessary now. It's just a wrapper for `stdin.close()` and I don't see it involved in any async work correct? If true just remove it imo and use `stdin.close()` instead

File: src/java/org/apache/cassandra/tools/JMXTool.java [hunk 22]
```diff
@@ -0,0 +1,838 @@
...
+        @Option(title = "format", name = { "-f", "--format" }, description = "What format the files are in")
+        private Format format = Format.yaml;
+
+        @Option(title = "ignore left", name = { "--ignore-missing-on-left" }, description = "Ignore results missing on the left")
+        private boolean ignoreMissingLeft;
+
+        @Option(title = "ignore right", name = { "--ignore-missing-on-right" }, description = "Ignore results missing on the right")
+        private boolean ignoreMissingRight;
+
+        @Option(title = "exclude objects", name = "--exclude-object", description = "Ignores processing specific objects")
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/tools/JMXTool.java]: Seems like a plural option from the description, but a single in the argument . Perhaps use `--exclude-objects` and more description on how to specify multiple objects. (Or if you specify the argument multiple times then remove the plural from description and indicate you can use multiple times.

File: src/java/org/apache/cassandra/tools/JMXTool.java [hunk 23]
```diff
@@ -0,0 +1,838 @@
...
+        private List<CliPattern> excludeObjects = new ArrayList<>();
+
+        @Option(title = "exclude attributes", name = "--exclude-attribute", description = "Ignores processing specific attributes")
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/tools/JMXTool.java]: Similar to `exclude-object`

File: src/java/org/apache/cassandra/tools/JMXTool.java [hunk 24]
```diff
@@ -0,0 +1,838 @@
...
+        private List<CliPattern> excludeAttributes = new ArrayList<>();
+
+        @Option(title = "exclude operations", name = "--exclude-operation", description = "Ignores processing specific operations")
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/tools/JMXTool.java]: Similar to `exclude-object`

File: src/java/org/apache/cassandra/tools/JMXTool.java [hunk 25]
```diff
@@ -0,0 +1,838 @@
...
+        {
+            MBeanServerConnection mbsc = jmxc.getMBeanServerConnection();
+
+            Map<String, Info> map = new TreeMap<>();
+            for (String pkg : new TreeSet<>(METRIC_PACKAGES))
+            {
+                Set<ObjectName> metricNames = new TreeSet<>(mbsc.queryNames(new ObjectName(pkg + ":*"), null));
+                for (ObjectName name : metricNames)
+                {
+                    if (mbsc.isRegistered(name))
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/tools/JMXTool.java:418]: Should you provide feedback to the user if the metric package is not registered?

File: src/java/org/apache/cassandra/tools/JMXTool.java [hunk 26]
```diff
@@ -0,0 +1,838 @@
...
+    private static Map<String, Info> load(JMXServiceURL url) throws IOException, MalformedObjectNameException, IntrospectionException, InstanceNotFoundException, ReflectionException
+    {
+        try (JMXConnector jmxc = JMXConnectorFactory.connect(url, null))
+        {
+            MBeanServerConnection mbsc = jmxc.getMBeanServerConnection();
+
+            Map<String, Info> map = new TreeMap<>();
+            for (String pkg : new TreeSet<>(METRIC_PACKAGES))
+            {
+                Set<ObjectName> metricNames = new TreeSet<>(mbsc.queryNames(new ObjectName(pkg + ":*"), null));
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/tools/JMXTool.java:415]: If the package contains no objects, should you provide feedback to the user to let them know?

File: src/java/org/apache/cassandra/tools/JMXTool.java [hunk 27]
```diff
@@ -0,0 +1,838 @@
...
+                access = "read-only";
+        }
+        else if (a.isWritable())
+            access = "write-only";
+        else
+            access = "no-access";
+        return access;
+    }
+
+    private static String normalizeType(String type)
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/tools/JMXTool.java:446]: Why do you only normalize arrays and not primitive types/objects too?

File: src/java/org/apache/cassandra/tools/JMXTool.java [hunk 28]
```diff
@@ -0,0 +1,838 @@
...
+                                           .toArray(Operation[]::new);
+            return new Info(attributes, operations);
+        }
+
+        public Attribute[] getAttributes()
+        {
+            return attributes;
+        }
+
+        public void setAttributes(Attribute[] attributes)
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/tools/JMXTool.java:517]: Is this here for completeness? Doesn't seem used.

File: src/java/org/apache/cassandra/tools/JMXTool.java [hunk 29]
```diff
@@ -0,0 +1,838 @@
...
+        {
+            return getAttribute(name).orElseThrow(AssertionError::new);
+        }
+
+        public Optional<Operation> getOperation(String name)
+        {
+            return Stream.of(operations).filter(o -> o.name.equals(name)).findFirst();
+        }
+
+        public Operation getOperationPresent(String name)
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/tools/JMXTool.java:567]: Couldn't find a use. Future proofing?

File: src/java/org/apache/cassandra/tools/JMXTool.java [hunk 30]
```diff
@@ -0,0 +1,838 @@
...
+                                           .toArray(Attribute[]::new);
+
+            Operation[] operations = Stream.of(info.getOperations())
+                                           .sorted(OPERATOR_COMPARATOR)
+                                           .map(Operation::from)
+                                           .toArray(Operation[]::new);
+            return new Info(attributes, operations);
+        }
+
+        public Attribute[] getAttributes()
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/tools/JMXTool.java:512]: Couldn't find a use. Future proofing?

File: src/java/org/apache/cassandra/tools/JMXTool.java [hunk 31]
```diff
@@ -0,0 +1,838 @@
...
+        {
+            return Stream.of(attributes).map(a -> a.name).collect(Collectors.toSet());
+        }
+
+        public Set<Attribute> attributeSet()
+        {
+            return new HashSet<>(Arrays.asList(attributes));
+        }
+
+        public Operation[] getOperations()
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/tools/JMXTool.java:532]: Couldn't find a use. Future proofing?

File: src/java/org/apache/cassandra/tools/JMXTool.java [hunk 32]
```diff
@@ -0,0 +1,838 @@
...
+        {
+            return operations;
+        }
+
+        public void setOperations(Operation[] operations)
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/tools/JMXTool.java:537]: Couldn't find a use. Future proofing?

File: src/java/org/apache/cassandra/tools/JMXTool.java [hunk 33]
```diff
@@ -0,0 +1,838 @@
...
+        {
+            this.operations = operations;
+        }
+
+        public Set<String> operationNames()
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/tools/JMXTool.java:542]: Couldn't find a use. Future proofing?

File: src/java/org/apache/cassandra/tools/JMXTool.java [hunk 34]
```diff
@@ -0,0 +1,838 @@
...
+        {
+            return new Attribute(info.getName(), normalizeType(info.getType()), JMXTool.getAccess(info));
+        }
+
+        public String getName()
+        {
+            return name;
+        }
+
+        public void setName(String name)
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/tools/JMXTool.java:618]: Couldn't find a use. Future proofing?

File: src/java/org/apache/cassandra/tools/JMXTool.java [hunk 35]
```diff
@@ -0,0 +1,838 @@
...
+    private static void printRow(PrintStream out, String... args)
+    {
+        ROW_BUFFER.setLength(0);
+        ROW_BUFFER.append("\t\t");
+        for (String a : args)
+            ROW_BUFFER.append(a).append("\t");
+        out.println(ROW_BUFFER);
+    }
+
+    public static final class Info
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/tools/JMXTool.java:483]: Are notifications considered part of this? The `StreamManager` exposes some.

Comment (jonmeredith) [inline src/java/org/apache/cassandra/tools/JMXTool.java:483]: Could this class and Attribute/Operation be immutable? Seems like it's constructed and then never altered.

File: test/unit/org/apache/cassandra/tools/JMXCompatabilityTest.java [hunk 36]
```diff
@@ -0,0 +1,178 @@
...
+                                                       "forceRepairRangeAsync", // -> repairAsync
+                                                       "beginLocalSampling", // -> beginLocalSampling(p1: java.lang.String, p2: int, p3: int): void
+                                                       "finishLocalSampling" // -> finishLocalSampling(p1: java.lang.String, p2: int): java.util.List
+        );
+
+        diff(excludeObjects, excludeAttributes, excludeOperations, "test/data/jmxdump/cassandra-3.0-jmx.yaml");
+    }
+
+    @Test
+    public void diff311() throws Throwable
```
Comment (jonmeredith) [inline test/unit/org/apache/cassandra/tools/JMXCompatabilityTest.java:105]: Interesting about whether to handle renames - `HintedHandOffManager` --> `HintsService`

File: tools/bin/jmxtool.bat [hunk 37]
```diff
@@ -0,0 +1,36 @@
...
+if "%OS%" == "Windows_NT" setlocal
+
+pushd "%~dp0"
+call cassandra.in.bat
+
+if NOT DEFINED JAVA_HOME goto :err
+
+set CASSANDRA_PARAMS=%CASSANDRA_PARAMS% -Dcassandra.logdir="%CASSANDRA_HOME%\logs"
+
+"%JAVA_HOME%\bin\java" -cp %CASSANDRA_CLASSPATH% %CASSANDRA_PARAMS% -Dlog4j.configurationFile=log4j2-tools.xml org.apache.cassandra.tools.JMXTool %*
```
Comment (jonmeredith) [inline tools/bin/jmxtool.bat]: log4j2 rather than logback

**Review by jonmeredith — COMMENTED** [review]
Only real bug is the log4j2/logback stuff. 
I was interested if you decided Notifications were out of scope.
The rest are usability things about the CLI and some questions on whether the JMX loader classes could be simplified a little.

File: src/java/org/apache/cassandra/io/util/DataOutputBuffer.java (hunk 11 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/io/util/DataOutputBuffer.java:93]: this behavior also matches other OutputStreams such as `java.io.ByteArrayOutputStream`, also it is an optional method in java

```
public void flush() throws IOException {
}
```

File: test/unit/org/apache/cassandra/tools/ToolRunner.java (hunk 20 above)
Comment (dcapwell) [reply test/unit/org/apache/cassandra/tools/ToolRunner.java:52]: used once in a method, but done each call; this caused a intellj warning so switched to this to make the class clear of warnings

File: tools/bin/jmxtool (hunk 13 above)
Comment (dcapwell) [reply tools/bin/jmxtool]: heh... I copy pasted from FQL...

File: src/java/org/apache/cassandra/tools/JMXTool.java (hunk 15 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/tools/JMXTool.java]: source is the best docs ever!... will fix

File: src/java/org/apache/cassandra/tools/JMXTool.java (hunk 22 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/tools/JMXTool.java]: > Or if you specify the argument multiple times then remove the plural from description and indicate you can use multiple times

yeah, works multiple times; will document.

File: src/java/org/apache/cassandra/tools/JMXTool.java (hunk 27 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/tools/JMXTool.java:446]: this is `Class.toString` which is harder to read for arrays, so did this so I could read the output without googling =D.

primitives and objects will normally have valid string names, but arrays will be like `J[` which require knowing what that means or googling it.

File: src/java/org/apache/cassandra/tools/JMXTool.java (hunk 28 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/tools/JMXTool.java:517]: snakeyaml and Jackson use the setters.

File: src/java/org/apache/cassandra/tools/JMXTool.java (hunk 29 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/tools/JMXTool.java:567]: this was when I implemented diff myself before I figured out I could add equality to the types to get set diff to support it... I think it is dead code now.

File: src/java/org/apache/cassandra/tools/JMXTool.java (hunk 30 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/tools/JMXTool.java:512]: Jackson/snakeyaml

File: src/java/org/apache/cassandra/tools/JMXTool.java (hunk 31 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/tools/JMXTool.java:532]: Jackson/snakeyaml

File: src/java/org/apache/cassandra/tools/JMXTool.java (hunk 32 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/tools/JMXTool.java:537]: Jackson/snakeyaml

File: src/java/org/apache/cassandra/tools/JMXTool.java (hunk 33 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/tools/JMXTool.java:542]: Jackson/snakeyaml

File: src/java/org/apache/cassandra/tools/JMXTool.java (hunk 34 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/tools/JMXTool.java:618]: Jackson/snakeyaml

File: tools/bin/jmxtool (hunk 13 above)
Comment (dcapwell) [reply tools/bin/jmxtool]:
```
$ grep 'log4j2-tools.xml' tools/bin/*
tools/bin/fqltool:        -Dlog4j.configurationFile=log4j2-tools.xml \
tools/bin/fqltool.bat:"%JAVA_HOME%\bin\java" -cp %CASSANDRA_CLASSPATH% %CASSANDRA_PARAMS% -Dlog4j.configurationFile=log4j2-tools.xml org.apache.cassandra.fqltool.FullQueryLogTool %*
tools/bin/jmxtool:        -Dlog4j.configurationFile=log4j2-tools.xml \
tools/bin/jmxtool.bat:"%JAVA_HOME%\bin\java" -cp %CASSANDRA_CLASSPATH% %CASSANDRA_PARAMS% -Dlog4j.configurationFile=log4j2-tools.xml org.apache.cassandra.tools.JMXTool %*
```

will fix all

File: src/java/org/apache/cassandra/tools/JMXTool.java (hunk 35 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/tools/JMXTool.java:483]: at the moment no, thought it simplest to add later; would be to just add that to info

Comment (dcapwell) [reply src/java/org/apache/cassandra/tools/JMXTool.java:483]: I would prefer immutable, but didn't know how to make snakeyaml happy with that, and would need to annotate each param for Jackson to support it =(

File: src/java/org/apache/cassandra/io/util/DataOutputBuffer.java (hunk 11 above)
Comment (jonmeredith) [reply src/java/org/apache/cassandra/io/util/DataOutputBuffer.java:93]: Fair enough, no objections.

File: test/unit/org/apache/cassandra/tools/JMXCompatabilityTest.java (hunk 36 above)
Comment (dcapwell) [reply test/unit/org/apache/cassandra/tools/JMXCompatabilityTest.java:105]: mixed feeling about this.  Right now I did what was needed to get these tests to pass; I am not saying that these changes are ok.

As far as I could tell, we wanted to move from HintedHandOffManager to HintsService a long time ago, but wasn't clear if all metrics are present in `HintsService` as 4.0 renamed and didn't have alias support for this code like we do for tables... We can always check what was present in `HintsService` (all gold files would show what's there for each version) to make sure nothing got dropped.

now, if we wanted to diff just those two that wouldn't be hard as it would be to just extract those two objects and diff them.


tested this real quick

```
@Test
    public void diff311HintService() throws Throwable
    {
        setupStandardTables();

        try (InputStream left = new FileInputStream("test/data/jmxdump/cassandra-3.11-jmx.yaml");
             InputStream right = new FileInputStream(TMP.getRoot().getAbsolutePath() + "/out.yaml"))
        {
            Map<String, JMXTool.Info> three11 = JMXTool.Diff.Format.yaml.load(left);
            Map<String, JMXTool.Info> trunk = JMXTool.Diff.Format.yaml.load(right);

            JMXTool.Info hintedHandOffManager = three11.get("org.apache.cassandra.db:type=HintedHandoffManager");
            JMXTool.Info hintService = trunk.get("org.apache.cassandra.hints:type=HintsService");

            JMXTool.Diff differ = new JMXTool.Diff();
            differ.diff(ImmutableMap.of("a", hintedHandOffManager), ImmutableMap.of("a", hintService));
        }
    }
```

```
Difference found in attribute or operation
a	operation not in right:
	deleteHintsForEndpoint(p1: java.lang.String): void
	listEndpointsPendingHints(): java.util.List
	pauseHintsDelivery(p1: boolean): void
	scheduleHintDelivery(p1: java.lang.String): void
	truncateAllHints(): void
a	operation not in left:
	deleteAllHints(): void
	deleteAllHintsForEndpoint(p1: java.lang.String): void
	pauseDispatch(): void
	resumeDispatch(): void

```

File: src/java/org/apache/cassandra/tools/JMXTool.java (hunk 26 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/tools/JMXTool.java:415]: I feel that consistency is all that matters in this tool, if it is empty in both clusters, that is all that matters

File: src/java/org/apache/cassandra/tools/JMXTool.java (hunk 25 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/tools/JMXTool.java:418]: I feel that consistency is all that matters in this tool, if it is not registered in both clusters, that is all that matters

**Review by jonmeredith — APPROVED** [review]
👍

File: test/unit/org/apache/cassandra/tools/ToolRunner.java (hunk 20 above)
Comment (bereng) [reply test/unit/org/apache/cassandra/tools/ToolRunner.java:52]: sgtm
